Dockerfile

```diff
@@ -1,12 +1,21 @@
-FROM alpine:edge
+# Build image
+FROM golang:alpine AS builder
 MAINTAINER Vishal Rana <vr@labstack.com>
 
-ENV VERSION 0.3.7
+COPY . /go/src/github.com/labstack/armor
```

Comment:
It doesn't work for me.
All external parts are missing:
```
Sending build context to Docker daemon  8.093MB
Step 1/10 : FROM golang:alpine AS builder
 ---> f421e93ece9c
Step 2/10 : MAINTAINER Vishal Rana <vr@labstack.com>
 ---> Using cache
 ---> b66c4ae6685a
Step 3/10 : COPY . /go/src/github.com/labstack/armor
 ---> 33cc5a3b5854
Step 4/10 : WORKDIR /go/src/github.com/labstack/armor
Removing intermediate container dc42cbcecb64
 ---> 87b84162cabd
Step 5/10 : RUN set -x     && export CGO_ENABLED=0     && go build -v -o /go/bin/armor cmd/armor/main.go
 ---> Running in a91cf64ab36b
+ export CGO_ENABLED=0
+ go build -v -o /go/bin/armor cmd/armor/main.go
plugin/plugin.go:10:2: cannot find package "github.com/Knetic/govaluate" in any of:
	/usr/local/go/src/github.com/Knetic/govaluate (from $GOROOT)
	/go/src/github.com/Knetic/govaluate (from $GOPATH)
cmd/armor/main.go:10:2: cannot find package "github.com/go-yaml/yaml" in any of:
	/usr/local/go/src/github.com/go-yaml/yaml (from $GOROOT)
	/go/src/github.com/go-yaml/yaml (from $GOPATH)
plugin/message.pb.go:16:8: cannot find package "github.com/golang/protobuf/proto" in any of:
	/usr/local/go/src/github.com/golang/protobuf/proto (from $GOROOT)
	/go/src/github.com/golang/protobuf/proto (from $GOPATH)
armor.go:7:2: cannot find package "github.com/labstack/echo" in any of:
	/usr/local/go/src/github.com/labstack/echo (from $GOROOT)
	/go/src/github.com/labstack/echo (from $GOPATH)
plugin/body_limit.go:5:2: cannot find package "github.com/labstack/echo/middleware" in any of:
	/usr/local/go/src/github.com/labstack/echo/middleware (from $GOROOT)
	/go/src/github.com/labstack/echo/middleware (from $GOPATH)
armor.go:8:2: cannot find package "github.com/labstack/gommon/color" in any of:
	/usr/local/go/src/github.com/labstack/gommon/color (from $GOROOT)
	/go/src/github.com/labstack/gommon/color (from $GOPATH)
armor.go:9:2: cannot find package "github.com/labstack/gommon/log" in any of:
	/usr/local/go/src/github.com/labstack/gommon/log (from $GOROOT)
	/go/src/github.com/labstack/gommon/log (from $GOPATH)
http/http.go:14:2: cannot find package "github.com/mitchellh/go-homedir" in any of:
	/usr/local/go/src/github.com/mitchellh/go-homedir (from $GOROOT)
	/go/src/github.com/mitchellh/go-homedir (from $GOPATH)
plugin/plugin.go:14:2: cannot find package "github.com/mitchellh/mapstructure" in any of:
	/usr/local/go/src/github.com/mitchellh/mapstructure (from $GOROOT)
	/go/src/github.com/mitchellh/mapstructure (from $GOPATH)
plugin/plugin.go:15:2: cannot find package "github.com/valyala/fasttemplate" in any of:
	/usr/local/go/src/github.com/valyala/fasttemplate (from $GOROOT)
	/go/src/github.com/valyala/fasttemplate (from $GOPATH)
http/http.go:15:2: cannot find package "golang.org/x/crypto/acme/autocert" in any of:
	/usr/local/go/src/golang.org/x/crypto/acme/autocert (from $GOROOT)
	/go/src/golang.org/x/crypto/acme/autocert (from $GOPATH)
The command '/bin/sh -c set -x     && export CGO_ENABLED=0     && go build -v -o /go/bin/armor cmd/armor/main.go' returned a non-zero code: 1
```

I would replace it by: `RUN go get -u github.com/labstack/armor`.
But this is just a guess.

Reply:
hm.. I'll try, maybe some changes was made in vendor

Reply:
> I can't run anything (mkdir) on your image.
> I maybe missing something...

If you need to do something like "mkdir" or other linux-tools, do that in your Dockerfile...

Reply:
I'm doing it in the Dockerfile of the next image and it doesn't work:

```
docker build -t armor-test .
Sending build context to Docker daemon  13.82kB
Step 1/7 : FROM armor-pull
 ---> 43c99df892a0
Step 2/7 : RUN mkdir -p /etc/armor/certs &&   mkdir -p /var/test
 ---> Running in e690865f20f2
OCI runtime create failed: container_linux.go:296: starting container process caused "exec: \"/bin/sh\": stat /bin/sh: no such file or directory": unknown
```

I maybe doing it wrong.

Reply:
I find a good workaround by using an intermediate container.

This looks good to me.

Reply:
it's because you copy `package.tar` to your container.. not extract, only copied

Reply:
You must do something like this:

```
# Build image
FROM golang:alpine AS builder
MAINTAINER Vishal Rana <vr@labstack.com>

COPY . /go/src/github.com/labstack/armor

WORKDIR /go/src/github.com/labstack/armor

RUN set -x \
    && export CGO_ENABLED=0 \
    && go build -v -o /go/bin/armor cmd/armor/main.go

RUN mkdir -p /etc/armor/certs

COPY ./config.yaml /etc/armor/config.yaml
COPY ./_fixture /var/www/test/

# Executable image
FROM scratch

WORKDIR /

COPY --from=builder /go/bin/armor /armor
COPY --from=builder /etc/armor /etc/armor
COPY --from=builder /var/www /var/www
COPY --from=builder /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/

ENTRYPOINT ["/armor", "-c", "/etc/armor/config.yaml"]
```

Reply:
Sorry, I'm just missing up...
This should work.

Reply:
nice.. waiting for review @vishr @alexandreStein @SkyrisBactera 🙂